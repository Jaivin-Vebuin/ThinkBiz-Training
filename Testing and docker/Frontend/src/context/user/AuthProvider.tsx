import { createContext, ReactNode, useEffect, useState } from "react";
import { useSelector } from "react-redux";
import { RootState } from "../../redux/store/store";
import { getUserAPI } from "../../services/userFetchServices";

type UserDetails = {
  id: string;
  name: string;
  email: string;
  role: string;
  age:number;
};

type AuthContextType = {
  isLoggedIn: boolean;
  currentUser: UserDetails | null;
};

type AuthProviderProps = {
  children: ReactNode;
};

const AuthProvider = ({ children }: AuthProviderProps) => {
  const token = useSelector((state: RootState) => state.auth.token);
  const [currentUser, setCurrentUser] = useState<UserDetails | null>(null);

  const isLoggedIn = token ? true : false;

  useEffect(() => {
    const fetchUserDetails = async () => {
      try {
        const userResponse = await getUserAPI();
        const data = userResponse?.data.message[0];
    
        setCurrentUser(data || null);
      } catch (error) {
        console.error("Error fetching current user details:", error);
      }
    };

    if (isLoggedIn) {
      fetchUserDetails();
    }
  }, [isLoggedIn, token]);

  return (
    <AuthContext.Provider value={{ isLoggedIn, currentUser }}>
      {children}
    </AuthContext.Provider>
  );
};

export const AuthContext = createContext<AuthContextType | undefined>(
  undefined
);

export { AuthProvider };
