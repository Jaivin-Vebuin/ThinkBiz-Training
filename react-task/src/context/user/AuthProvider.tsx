import { createContext, useEffect, useState } from "react";
import { useSelector } from "react-redux";
import { RootState } from "../../redux/store/store";
import { getUserAPI } from "../../services/userFetchServices";
import { AuthContextType, AuthProviderProps, UserDetails } from "../../data/model/types/authProvider/authProviderTypes";

const AuthProvider = ({ children }: AuthProviderProps) => {
  const token = useSelector((state: RootState) => state.auth.token);
  const [currentUser, setCurrentUser] = useState<UserDetails | null>(null);

  const isLoggedIn = token ? true : false;

  useEffect(() => {
    const fetchUserDetails = async () => {
      try {
        const userResponse = await getUserAPI();
        const data = userResponse?.data.message[0];
    
        setCurrentUser(data || null);
      } catch (error) {
        console.error("Error fetching current user details:", error);
      }
    };

    if (isLoggedIn) {
      fetchUserDetails();
    }
  }, [isLoggedIn, token]);

  return (
    <AuthContext.Provider value={{ isLoggedIn, currentUser }}>
      {children}
    </AuthContext.Provider>
  );
};

export const AuthContext = createContext<AuthContextType | undefined>(
  undefined
);

export { AuthProvider };
