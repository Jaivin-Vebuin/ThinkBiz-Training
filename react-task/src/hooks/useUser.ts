import { useContext } from "react";
import { AuthContext } from "../context/user/AuthProvider";

const useUser = () => {
  const context = useContext(AuthContext);
  if (!context) {
    throw new Error("useUser must be used within an AuthProvider");
  }
  return context;
};
export default useUser;
