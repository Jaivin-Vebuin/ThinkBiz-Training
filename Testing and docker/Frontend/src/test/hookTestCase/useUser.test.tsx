import { renderHook } from "@testing-library/react";
import { AuthContext } from "../../context/user/AuthProvider";
import useUser from "../../hooks/useUser";
import { ReactNode } from "react";

const mockAuthContext = {
  isLoggedIn: true,
  currentUser: {
    id: "123",
    name: "John Doe",
    email: "john.doe@example.com",
    role: "admin",
    age: 30,
  },
};

describe("useUser Hook", () => {
  test("returns context value when inside AuthProvider", () => {
    const wrapper = ({ children }: { children: ReactNode }) => (
      <AuthContext.Provider value={mockAuthContext}>{children}</AuthContext.Provider>
    );

    const { result } = renderHook(() => useUser(), { wrapper });

    expect(result.current).toEqual(mockAuthContext);
  });

  test("throws an error when used outside AuthProvider", () => {
    const consoleErrorMock = jest.spyOn(console, "error").mockImplementation(() => {});

    expect(() => renderHook(() => useUser())).toThrow(
      "useUser must be used within an AuthProvider"
    );

    consoleErrorMock.mockRestore();
  });

  test("matches the snapshot", () => {
    const wrapper = ({ children }: { children: ReactNode }) => (
      <AuthContext.Provider value={mockAuthContext}>{children}</AuthContext.Provider>
    );

    const { result } = renderHook(() => useUser(), { wrapper });

    expect(result.current).toMatchSnapshot();
  });
});
