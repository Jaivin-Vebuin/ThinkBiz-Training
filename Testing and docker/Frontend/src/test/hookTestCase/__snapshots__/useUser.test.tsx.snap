// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useUser Hook matches the snapshot 1`] = `
{
  "currentUser": {
    "age": 30,
    "email": "john.doe@example.com",
    "id": "123",
    "name": "John Doe",
    "role": "admin",
  },
  "isLoggedIn": true,
}
`;
